src/main/java/pl/cleankod/exchange/provider/FindAccountService.java

```diff
@@ -0,0 +1,48 @@
+package pl.cleankod.exchange.provider;
+
+import pl.cleankod.exchange.core.domain.Account;
+import pl.cleankod.exchange.core.usecase.FindAccountAndConvertCurrencyUseCase;
+import pl.cleankod.exchange.core.usecase.FindAccountUseCase;
+import pl.cleankod.util.AccountFailedReason;
+import pl.cleankod.util.Preconditions;
+import pl.cleankod.util.Result;
+
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Currency;
+import java.util.Optional;
+
+public class FindAccountService {
+    private final FindAccountAndConvertCurrencyUseCase findAccountAndConvertCurrencyUseCase;
+    private final FindAccountUseCase findAccountUseCase;
+
+    public FindAccountService(
+            FindAccountAndConvertCurrencyUseCase findAccountAndConvertCurrencyUseCase,
+            FindAccountUseCase findAccountUseCase
+    ) {
+        this.findAccountAndConvertCurrencyUseCase = findAccountAndConvertCurrencyUseCase;
+        this.findAccountUseCase = findAccountUseCase;
+    }
+
+    public Result<Account, AccountFailedReason> findAccountById(String id, String currency) {
+        Preconditions.requireNonNull(id);
+
+        Account.Id accountId = Account.Id.of(URLDecoder.decode(id, StandardCharsets.UTF_8));
```

Comment:
The web layer bled to another one. Why not accept the domain object as parameters at this point already? Now this class breaks the Single-responsibility principle, because it also takes care of the value conversions.